src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,433 @@
+{
+	Size size = 0;
+
+	size = add_size(size, sizeof(DistributedTransactionShmemData));
+	size = add_size(size, mul_size(sizeof(DistributedTransactionBackendData),
+								   MaxBackends));
+
+	return size;
+}
+
+
+/*
+ *  InitializeDistributedTransactionManagementBackend is called per backend
+ *  and does the required initilizations.
+ */
+void
+InitializeDistributedTransactionManagementBackend(void)
+{
+	LWLockAcquire(&distributedTransactionShmemData->lock, LW_EXCLUSIVE);
+
+	DistributedTransactionBackend =
+		&distributedTransactionShmemData->sessions[MyProc->pgprocno];
+
+	Assert(DistributedTransactionBackend);
+
+	DistributedTransactionBackend->databaseId = 0;
+	DistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+	DistributedTransactionBackend->transactionId.transactionId = 0;
+	DistributedTransactionBackend->transactionId.timestamp = 0;
+
+	/* we'll use mutex while accessing the backend shmem */
+	SpinLockInit(&BackendDistributedTransactionMutex);
+
+	LWLockRelease(&distributedTransactionShmemData->lock);
+}
+
+
+/*
+ * UnSetDistributedTransactionId simply acquires the mutex and resets
+ * the backend's distributed transaction data in shared memory to the
+ * initial values.
+ */
+void
+UnSetDistributedTransactionId(void)
+{
+	if (DistributedTransactionBackend)
+	{
+		SpinLockAcquire(&BackendDistributedTransactionMutex);
+
+		DistributedTransactionBackend->databaseId = 0;
+		DistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+		DistributedTransactionBackend->transactionId.transactionId = 0;
+		DistributedTransactionBackend->transactionId.timestamp = 0;
+
+		SpinLockRelease(&BackendDistributedTransactionMutex);
+	}
+}
+
+
+/*
+ * GenerateNextDistributedTransactionId returns a new distributed transaction id with
+ * the current timestamp and next local transaction id.
+ */
+DistributedTransactionId *
+GenerateNextDistributedTransactionId(void)
+{
+	DistributedTransactionId *nextDistributedTransactionId =
+		(DistributedTransactionId *) palloc(sizeof(DistributedTransactionId));
+
+	/*
+	 * FIXME: Once we allow running queries from secondaries, we should generate
+	 * an id which is unique to the node, not to the group as we do here.
+	 */
+	nextDistributedTransactionId->initiatorNodeIdentifier = (uint64) GetLocalGroupId();
+
+	nextDistributedTransactionId->transactionId = GetNextLocalTransactionIdFromShmem();
+	nextDistributedTransactionId->timestamp = GetCurrentTimestamp();
+
+	return nextDistributedTransactionId;
+}
+
+
+/*
+ * GetNextLocalTransactionIdFromShmem atomically fetches and returns
+ * the next distributed transaction id from the shared memory.
+ */
+static uint64
+GetNextLocalTransactionIdFromShmem(void)
+{
+	pg_atomic_uint64 *transactionIdSequence =
+		&distributedTransactionShmemData->nextTransactionId;
+	uint64 nextTransactionId = 0;
+
+	nextTransactionId = pg_atomic_fetch_add_u64(transactionIdSequence, 1);
+
+	return nextTransactionId;
```

Comment:
Personally this'd be a lot easier to read if you'd just skip all the local variables - they don't help...